### PR TITLE
[WIP] C# 8 nullable reference type support 

### LINKED ISSUE #54: Feature request: C# 8 nullable reference type support
It would be great if PublicApiGenerator "understood" NullableAttribute, so could print out the API with suitable `?` syntax.
Trickier: exposing whether an API is null-aware at all.

More info: https://codeblog.jonskeet.uk/2019/02/10/nullableattribute-and-c-8/
Sample code that's way hackier than PublicApiGenerator: https://github.com/nodatime/nodatime/tree/master/build/NullableApiReviewer

### PR DESCRIPTION
**Pull request by sungam3r**
Rebased #99 

I did not understand how to do rebase right, the branch was deleted and the previous PR closed, I had to create a new one.

fixes #54

### CODE REVIEW

File: src/PublicApiGeneratorTests/ApiGeneratorTestsBase.cs [hunk 1]
```diff
@@ -19,9 +19,9 @@ protected void AssertPublicApi(Type type, string expectedOutput, bool includeAss
             AssertPublicApi(new[] { type }, expectedOutput, includeAssemblyAttributes, excludeAttributes: excludeAttributes);
         }
 
-        protected void AssertPublicApi(Type[] types, string expectedOutput, bool includeAssemblyAttributes = false, string[] whitelistedNamespacePrefixes = default(string[]), string[] excludeAttributes = null)
+        protected void AssertPublicApi(Type[] types, string expectedOutput, bool includeAssemblyAttributes = false, string[] whitelistedNamespacePrefixes = default, string[] excludeAttributes = null)
         {
-            AssertPublicApi(GetType().Assembly, types, expectedOutput, includeAssemblyAttributes, whitelistedNamespacePrefixes, excludeAttributes);
+            AssertPublicApi(types[0].Assembly, types, expectedOutput, includeAssemblyAttributes, whitelistedNamespacePrefixes, excludeAttributes);
```
Comment (sungam3r) [inline src/PublicApiGeneratorTests/ApiGeneratorTestsBase.cs]: @danielmarbach It seems there was a mistake if I use type from external assembly.

Comment (danielmarbach) [reply src/PublicApiGeneratorTests/ApiGeneratorTestsBase.cs]: Ok. No worries I can have another look before merging. Ping me when I should do a final review

Comment (sungam3r) [reply src/PublicApiGeneratorTests/ApiGeneratorTestsBase.cs]: PR is ready for review.

File: src/PublicApiGeneratorTests/NullableTests.cs [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+        }
+
+        [Fact]
+        public void Should_Annotate_OpenGeneric()
+        {
+            AssertPublicApi(typeof(StringNullableList<>),
+@"namespace PublicApiGeneratorTests.Examples
+{
+    public class StringNullableList<T> : System.Collections.Generic.List<T?>
+        where T : struct
```
Comment (jnm2) [inline src/PublicApiGeneratorTests/NullableTests.cs]: Isn't the `where T : class` case the interesting one?

Comment (sungam3r) [reply src/PublicApiGeneratorTests/NullableTests.cs]: done

**Comment by danielmarbach** [comment]
I don't think it is a good idea to bundle so many things into the same PR. Let's keep this one focused. I'm ok to take this one with 107 and 108 but let's not add more

**Comment by danielmarbach** [comment]
Is this still WIP?

**Comment by sungam3r** [comment]
I did not see your messages. Yes, it was at work, but I see that it is time to finish. I rolled back the last commit related to `dynamic` support.

**Comment by sungam3r** [comment]
@danielmarbach @jnm2 I made a new (final) PR #115 with squashed commits.

**Comment by jnm2** [comment]
You can always force-push instead of closing the PR and starting a new one. GitHub displays an event describing the force-push with a link to compare before and after.

**Pull request closed without merging**